### PR TITLE
Support optimizer step progress tracking with manual optimization

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Adds optimizer step progress tracking support to manual optimization.

Part of #11805

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @awaelchli

### CODE REVIEW

File: tests/trainer/optimization/test_manual_optimization.py [hunk 1]
```diff
@@ -553,115 +550,50 @@ def configure_optimizers(self):
...
         log_every_n_steps=1,
     )
 
     with mock.patch.object(Strategy, "backward", wraps=trainer.strategy.backward) as bwd_mock:
         trainer.fit(model)
     assert bwd_mock.call_count == limit_train_batches * 2
-
-
-@patch("torch.optim.SGD.step")
-def test_step_with_optimizer_closure_and_extra_arguments(step_mock, tmpdir):
```
Comment (carmocca) [inline tests/trainer/optimization/test_manual_optimization.py:611]: I believe this test is duplicated

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2009,6 +2009,7 @@ def data_parallel_device_ids(self) -> Optional[List[int]]:
 
     @property
     def lightning_module(self) -> "pl.LightningModule":
+        # TODO: this is actually an optional return
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2012]: Not planning to resolve this here but I wanted to point out that the annotation is incorrect.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -333,7 +333,7 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
         checkpoint = {
             # the epoch is saved for compatibility but it's not relevant for restoration
             "epoch": self.trainer.current_epoch,
-            "global_step": self.trainer.global_step + 1,
+            "global_step": self.trainer.global_step + model.automatic_optimization,
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:336]: The plan is to fix it for manual also in the follow-up: #11805

File: pytorch_lightning/loops/optimization/manual_loop.py [hunk 4]
```diff
@@ -74,6 +76,10 @@ class ManualOptimization(Loop[_OUTPUTS_TYPE]):
 
     def __init__(self) -> None:
         super().__init__()
+        # since manual optimization does not track lr scheduler or optimizer frequencies, we use a simpler progress than
+        # `OptimizationProgress`
+        self.optim_step_progress = Progress.from_defaults(ReadyCompletedTracker)
```
Comment (ethanwharris) [inline pytorch_lightning/loops/optimization/manual_loop.py:81]: Personal preference but I would go for full name rather than contraction. E.g.:
```suggestion
        self.optimizer_step_progress = Progress.from_defaults(ReadyCompletedTracker)
```

**Review by ethanwharris — APPROVED** [review]
LGTM, minor comment

File: tests/trainer/optimization/test_manual_optimization.py [hunk 5]
```diff
@@ -167,6 +167,7 @@ def training_epoch_end(self, outputs) -> None:
     with mock.patch.object(Strategy, "backward", wraps=trainer.strategy.backward) as bwd_mock:
         trainer.fit(model)
     assert bwd_mock.call_count == limit_train_batches * 3
+    assert trainer.global_step == limit_train_batches * 2
```
Comment (awaelchli) [inline tests/trainer/optimization/test_manual_optimization.py:170]: the model has two optimizers and we call both of them, so this value makes sense. I'm not sure though which change in the loops is causing this / why it was 3 before? Is it because of fit_loop line 77?

File: tests/trainer/optimization/test_manual_optimization.py (hunk 1 above)
Comment (awaelchli) [reply tests/trainer/optimization/test_manual_optimization.py:611]: where did you see it?

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: tests/trainer/optimization/test_manual_optimization.py (hunk 1 above)
Comment (carmocca) [reply tests/trainer/optimization/test_manual_optimization.py:611]: The one right above.

And the assertion is covered by `test_step_with_optimizer_closure_with_different_frequencies`

File: tests/trainer/optimization/test_manual_optimization.py (hunk 5 above)
Comment (carmocca) [reply tests/trainer/optimization/test_manual_optimization.py:170]: It wasn't 3 before, we didn't have tests for it but it would have been `== limit_train_batches`
